File: 0076-minimum-window-substring/0076-minimum-window-substring.py
from collections import Counter

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not t or not s or len(s) < len(t):
            return ""

        # Count the frequencies of characters we need from t
        need = Counter(t)
        window = {}

        # 'have' is how many unique characters in our window meet the required frequency
        # 'required' is the total number of unique characters in t
        have, required = 0, len(need)
        
        # Pointers for our sliding window
        l, r = 0, 0
        
        # To store the result: [window_length, left_bound, right_bound]
        res = [float('infinity'), -1, -1]

        while r < len(s):
            # Add the right character to our window
            char = s[r]
            window[char] = window.get(char, 0) + 1

            # If the character is in t and we have the exact amount needed, increment 'have'
            if char in need and window[char] == need[char]:
                have += 1

            # When our window contains all required characters, try to shrink it
            while have == required:
                # Update our minimum window result
                if (r - l + 1) < res[0]:
                    res = [r - l + 1, l, r]

                # Remove the left character from our window
                left_char = s[l]
                window[left_char] -= 1

                # If removing it breaks our valid window, decrement 'have'
                if left_char in need and window[left_char] < need[left_char]:
                    have -= 1
                
                # Move the left pointer forward to shrink
                l += 1
            
            # Move the right pointer forward to expand
            r += 1

        # If we never found a valid window, return an empty string
        l, r = res[1], res[2]
        return s[l:r+1] if res[0] != float('infinity') else ""

File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        if (s.empty() || t.empty() || s.length() < t.length()) {
            return "";
        }

        // Count the frequencies of characters we need from t
        unordered_map<char, int> need;
        for (char c : t) {
            need[c]++;
        }

        unordered_map<char, int> window;
        
        // 'have' is how many unique characters in our window meet the required frequency
        // 'required' is the total number of unique characters in t
        int have = 0;
        int required = need.size();
        
        // Pointers for our sliding window
        int l = 0;
        int r = 0;
        
        // To store the result: min_len (infinity equivalent), and start index
        int min_len = -1; // We use -1 to represent infinity
        int min_start = 0;

        while (r < s.length()) {
            // Add the right character to our window
            char c = s[r];
            window[c]++;

            // If the character is in t and we have the exact amount needed, increment 'have'
            // need.count(c) ensures we only care about characters actually in 't'
            if (need.count(c) && window[c] == need[c]) {
                have++;
            }

            // When our window contains all required characters, try to shrink it
            while (have == required) {
                // Update our minimum window result
                if (min_len == -1 || (r - l + 1) < min_len) {
                    min_len = r - l + 1;
                    min_start = l;
                }

                // Remove the left character from our window
                char left_char = s[l];
                window[left_char]--;

                // If removing it breaks our valid window, decrement 'have'
                if (need.count(left_char) && window[left_char] < need[left_char]) {
                    have--;
                }
                
                // Move the left pointer forward to shrink
                l++;
            }
            
            // Move the right pointer forward to expand
            r++;
        }

        // If min_len was never updated, return an empty string. 
        // Otherwise, extract the substring.
        return min_len == -1 ? "" : s.substr(min_start, min_len);
    }
};